core/lb.go

```diff
@@ -0,0 +1,202 @@
+package core
+
+import (
+	"errors"
+	"math"
+	"strings"
+	"sync"
+
+	"github.com/golang/glog"
+
+	"github.com/livepeer/go-livepeer/common"
+	"github.com/livepeer/lpms/ffmpeg"
+)
+
+var ErrTranscoderBusy = errors.New("TranscoderBusy")
+
+type TranscoderSession interface {
+	Transcoder
+	Stop()
+}
+
+type newTranscoderFn func(device string, workDir string) TranscoderSession
+
+type LoadBalancingTranscoder struct {
+	transcoders []string // Slice of device IDs
+	workDir     string
+	newT        newTranscoderFn
+
+	// The following fields need to be protected by the mutex `mu`
+	mu       *sync.RWMutex
+	load     map[string]int
+	sessions map[string]*transcoderSession
+	idx      int // Ensures a non-tapered work distribution
+}
+
+func NewLoadBalancingTranscoder(devices string, workDir string, newTranscoderFn newTranscoderFn) Transcoder {
+	d := strings.Split(devices, ",")
+	return &LoadBalancingTranscoder{
+		transcoders: d,
+		workDir:     workDir,
+		newT:        newTranscoderFn,
+		mu:          &sync.RWMutex{},
+		load:        make(map[string]int),
+		sessions:    make(map[string]*transcoderSession),
+	}
+}
+
+func (lb *LoadBalancingTranscoder) Transcode(job string, fname string, profiles []ffmpeg.VideoProfile) (*TranscodeData, error) {
+
+	lb.mu.RLock()
+	session, exists := lb.sessions[job]
+	lb.mu.RUnlock()
+	if exists {
+		glog.V(common.DEBUG).Info("LB: Using existing transcode session for ", session.key)
+	} else {
+		var err error
+		session, err = lb.createSession(job, fname, profiles)
+		if err != nil {
+			return nil, err
+		}
+	}
+	return session.Transcode(job, fname, profiles)
+}
+
+func (lb *LoadBalancingTranscoder) createSession(job string, fname string, profiles []ffmpeg.VideoProfile) (*transcoderSession, error) {
+
+	lb.mu.Lock()
+	defer lb.mu.Unlock()
+	glog.V(common.DEBUG).Info("LB: Creating transcode session for ", job)
+	transcoder := lb.leastLoaded()
+
+	// Acquire transcode session. Map to job id + assigned transcoder
+	key := job + "_" + transcoder
+	nonce := common.RandName() // because key alone is prone to reuse
+	costEstimate := calculateCost(profiles)
+	session := &transcoderSession{
+		transcoder: lb.newT(transcoder, lb.workDir),
+		key:        key,
+		nonce:      nonce,
+		sender:     make(chan *transcoderParams, 1),
+	}
+	lb.sessions[job] = session
+	lb.load[transcoder] += costEstimate
+	lb.idx = (lb.idx + 1) % len(lb.transcoders)
+
+	// Local cleanup function
+	cleanupSession := func() {
+		lb.mu.Lock()
+		defer lb.mu.Unlock()
+		sess, exists := lb.sessions[job]
+		if !exists {
+			return
+		}
+		if sess.nonce != session.nonce {
+			// Session cached at `job` has since been recreated
```

Comment:
How can this happen? Only way I see is the problem in Transcode function I've commented on earlier.
And when it happens, this will leave `lb.load` not decremented?

Reply:
>  this will leave lb.load not decremented?

Yes, good point.

Reply:
Fixed in https://github.com/livepeer/go-livepeer/pull/1124/commits/b9e09ffeac9470c420a0c489cb4df12f4bb2df8d